### DISCUSSION TITLE
应用侧出现Unknown system variable 'tx_read_only'报错

### DISCUSSION THREAD

**Discussion by wenyh1** [actiontech/dble#3067] [Q&A] [answered]
DBLE版本3.21.02.0，使用读写分离，fakeMysqlversion配置为5.7.11
DBLE后端MySQL版本为8.0
应用使用JDBC连接dble后，连接dble会报“Unknown system variable 'tx_read_only'”

**Comment by PanternBao** [comment, marked as answer]
# 背景
涉及的MYSQL的参数：

- tx_read_only (用于设置事务的只读模式)
- tx_isolation (用于设置事务的隔离级别)

> transaction_read_only与transaction_isolation参数在MYSQL5.7.20 版本引入，tx_read_only与tx_isolation参数在 MYSQL8.0 版本被移除。[见MYSQL官网](https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html)

### 报错原因
JDBC会根据获取DBLE版本信息fakeMysqlversion，决定是下发transaction_read_only还是tx_read_only。[见JDBC源码](https://github.com/mysql/mysql-connector-j/blob/e920b979015ae7117d60d72bcc8f077a839cd791/src/main/user-impl/java/com/mysql/cj/jdbc/ConnectionImpl.java#L1392)

- \>=5.7.20的版本会下发 transaction_*参数
- 其他版本会下发 tx_*参数

若下发的参数和后端MySQL不兼容时，会产生报错。

### 解决方式
在DBLE中的boostrap.cnf的fakeMysqlversion设置为\>=5.7.20版本

**Reply by yanhuqing666** [reply]
如果后侧数据库混用会怎么样？读写分离模式呢？

**Reply by wenyh1** [reply]
> 如果后侧数据库混用会怎么样？读写分离模式呢？

此场景，举例：
如，rwSplitUser1对应MYSQL5.7.11，rwSplitUser2对应MYSQL8.0；fakeMysqlversion设置为<5.7.20时

JDBC连接DBLE，现象如下：
使用rwSplitUser1用户连接，运行正常
使用rwSplitUser2用户连接时，出现“Unknown system variable 'tx_read_only'”报错

因此，DBLE中不推荐后端实例MYSQL大版本(5.7和8.0)混用使用

**Reply by yanhuqing666** [reply]
也就是说，读写分离模式也受到此不能混用的限制，理论上不应该，未来需要额外设计
